File: clients/java/test/java/com/hook0/client/JsonTest.java
package com.hook0.client;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.Timeout;
import org.junit.jupiter.api.function.Executable;

/**
 * The reader and writer this artefact carries instead of a dependency.
 *
 * <p>What is asserted is what a bounded reader has to do: read the subset the API speaks exactly, and <em>refuse</em>
 * at every ceiling rather than trim to it. Half a document read as a whole one is worse than no document at all.
 */
@Timeout(60)
final class JsonTest {

  @Test
  void aDocumentReadsAsTheValuesItCarries() {
    Object read = Json.parse("{\"a\":1,\"b\":[true,null,\"x\"],\"c\":{\"d\":1.5}}");

    Map<String, Object> members = Wire.asFields(read, "a document");

    assertEquals(Long.valueOf(1), members.get("a"));
    assertEquals(List.of(Boolean.TRUE, "x"), List.of(((List<?>) members.get("b")).get(0),
        ((List<?>) members.get("b")).get(2)));
    assertEquals(Double.valueOf(1.5), Wire.asFields(members.get("c"), "c").get("d"));
  }

  @Test
  void aWholeNumberReadsAsOneAndAFractionalOneDoesNot() {
    assertEquals(Long.valueOf(-42), Json.parse("-42"));
    assertEquals(Double.valueOf(1e3), Json.parse("1e3"));
    assertEquals(Double.valueOf(0.5), Json.parse("0.5"));
  }

  @Test
  void everythingAReaderHasToRefuseIsRefused() {
    for (String malformed :
        List.of("", "{", "[1,", "{\"a\"}", "{\"a\":}", "tru", "\"unterminated", "1 2", "{\"a\":1}x", "\\", "'a'")) {
      assertThrows(JsonException.class, () -> Json.parse(malformed), "`" + malformed + "` was read as a document");
    }
  }

  @Test
  void aDocumentNestedDeeperThanTheCeilingIsRefusedRatherThanTrimmed() {
    String tower = "[".repeat(Json.MAX_DEPTH + 4) + "1" + "]".repeat(Json.MAX_DEPTH + 4);

    JsonException refused = assertThrows(JsonException.class, () -> Json.parse(tower));

    assertTrue(refused.getMessage().contains("nests deeper"), refused.getMessage());
  }

  @Test
  void aDocumentLongerThanTheCeilingIsRefusedRatherThanTrimmed() {
    String oversized = "\"" + "x".repeat(Json.MAX_DOCUMENT_CHARS) + "\"";

    JsonException refused = assertThrows(JsonException.class, () -> Json.parse(oversized));

    assertTrue(refused.getMessage().contains("characters long"), refused.getMessage());
  }

  @Test
  void aStringLongerThanTheCeilingIsRefusedRatherThanTrimmed() {
    // Inside the document ceiling and past the string one, so what refuses it is the bound this case
    // is about.
    String oversized = "[\"" + "x".repeat(Json.MAX_STRING_CHARS + 8) + "\"]";

    JsonException refused = assertThrows(JsonException.class, () -> Json.parse(oversized));

    assertTrue(refused.getMessage().contains("characters read"), refused.getMessage());
  }

  @Test
  void whatIsWrittenReadsBackAsWhatWasWritten() {
    Map<String, Object> value = new LinkedHashMap<>();
    value.put("text", "a \"quoted\" \\ line\nand a tab\t");
    value.put("whole", Long.valueOf(7));
    value.put("fractional", Double.valueOf(0.25));
    value.put("flag", Boolean.FALSE);
    value.put("nothing", null);
    value.put("items", List.of("a", "b"));

    assertEquals(value, Json.parse(Json.write(value)));
  }

  @Test
  void aControlCharacterTravelsAsTheCodePointItIs() {
    String carried = "a" + ((char) 1) + "b";

    assertEquals("\"a\\u0001b\"", Json.write(carried));
    assertEquals(carried, Json.parse("\"a\\u0001b\""));
  }

  @Test
  void aValueJsonCannotCarryIsRefusedRatherThanPrinted() {
    JsonException refused = assertThrows(JsonException.class, () -> Json.write(new Object()));

    assertTrue(refused.getMessage().contains("not something JSON carries"), refused.getMessage());
    assertThrows(JsonException.class, () -> Json.write(Double.valueOf(Double.NaN)));
    assertThrows(JsonException.class, () -> Json.write(Map.of(Integer.valueOf(1), "keyed by a number")));
  }

  @Test
  void everyCeilingThisWriterNamesRefusesOnTheWayOutAsWellAsOnTheWayIn() {
    // Each bound is written down twice, once for reading and once for writing, and until now only
    // the reading half was held to it. A ceiling that is only enforced in one direction is a client
    // that will happily compose a document it would refuse to read back.
    Map<String, Object> members = new HashMap<>();
    for (int member = 0; member <= Json.MAX_ITEMS; member++) {
      members.put("m" + member, Long.valueOf(member));
    }
    List<Object> items = new ArrayList<>(Json.MAX_ITEMS + 1);
    for (int item = 0; item <= Json.MAX_ITEMS; item++) {
      items.add(Long.valueOf(item));
    }
    Object nested = Long.valueOf(1);
    for (int depth = 0; depth <= Json.MAX_DEPTH; depth++) {
      nested = List.of(nested);
    }
    Object tower = nested;
    List<Object> whole = new ArrayList<>();
    for (int part = 0; part < 10; part++) {
      whole.add("x".repeat(900_000));
    }

    assertTrue(refusing(() -> Json.write(members)).contains("members, above"));
    assertTrue(refusing(() -> Json.write(items)).contains("items, above"));
    assertTrue(refusing(() -> Json.write(tower)).contains("nests deeper"));
    assertTrue(refusing(() -> Json.write("x".repeat(Json.MAX_STRING_CHARS + 8))).contains("characters long"));
    assertTrue(refusing(() -> Json.write(whole)).contains("characters long, above"));

    // And the reading half of the two collection ceilings, which the cases above hold only the
    // writer to. A document carrying more than this many is refused where it crosses, rather than
    // read up to the ceiling and handed over as though that were all the sender wrote.
    StringBuilder crowded = new StringBuilder("{");
    StringBuilder crowdedList = new StringBuilder("[");
    for (int item = 0; item <= Json.MAX_ITEMS; item++) {
      crowded.append(item > 0 ? "," : "").append("\"m").append(item).append("\":1");
      crowdedList.append(item > 0 ? "," : "").append('1');
    }

    assertTrue(refusing(() -> Json.parse(crowded + "}")).contains("members read"));
    assertTrue(refusing(() -> Json.parse(crowdedList + "]")).contains("items read"));
  }

  @Test
  void everyCharacterThisWriterEscapesIsOneTheReaderTakesBack() {
    // The escape table is the one place where writing and reading have to agree character for
    // character: a writer that escapes what the reader does not unescape produces documents only it
    // can read. Each is asserted as the text that travels, then read back as the character it was.
    Map<String, String> escaped = new LinkedHashMap<>();
    escaped.put("\"", "\\\"");
    escaped.put("\\", "\\\\");
    escaped.put("\n", "\\n");
    escaped.put("\r", "\\r");
    escaped.put("\t", "\\t");
    escaped.put("\b", "\\b");
    escaped.put("\f", "\\f");

    for (Map.Entry<String, String> pair : escaped.entrySet()) {
      assertEquals("\"" + pair.getValue() + "\"", Json.write(pair.getKey()), "wrote `" + pair.getKey() + "`");
      assertEquals(pair.getKey(), Json.parse("\"" + pair.getValue() + "\""), "read `" + pair.getValue() + "`");
    }

    // A solidus is one JSON allows escaped and this writer leaves alone, so the reader has to take
    // it either way — documents written elsewhere carry it.
    assertEquals("/", Json.parse("\"\\/\""));
    assertEquals("\"/\"", Json.write("/"));
  }

  @Test
  void aStringThatStopsInsideAnEscapeIsRefusedRatherThanReadUpToThere() {
    // Four ways an escape can be cut short or wrong. Each of them ends a string somewhere the reader
    // cannot know what it was, and reading what came before it as the whole string would hand a
    // caller a value the sender never wrote.
    for (String malformed : List.of("\"a\\", "\"a\\u00\"", "\"a\\uzzzz\"", "\"a\\q\"")) {
      assertThrows(JsonException.class, () -> Json.parse(malformed), "`" + malformed + "` was read as a string");
    }
  }

  @Test
  void anObjectThatRunsOnWithoutASeparatorIsRefusedRatherThanReadUpToThere() {
    for (String malformed : List.of("{\"a\":1 \"b\":2}", "[1 2]")) {
      assertThrows(JsonException.class, () -> Json.parse(malformed), "`" + malformed + "` was read as a document");
    }
  }

  @Test
  void aNumberWithNothingAfterThePointIsWrittenWithoutOne() {
    // What the API reads is JSON, where `7` and `7.0` are the same number, and what a caller handed
    // in was a floating-point zero-fraction rather than a whole number. It travels as the shorter of
    // the two, from either width of float, and comes back as the whole number it now is.
    assertEquals("7", Json.write(Double.valueOf(7)));
    assertEquals("7", Json.write(Float.valueOf(7)));
    assertEquals("-7", Json.write(Double.valueOf(-7)));
    assertEquals("1.5", Json.write(Float.valueOf(1.5f)));
    assertEquals(Long.valueOf(7), Json.parse(Json.write(Double.valueOf(7))));

    // Past where a double can tell whole numbers apart one from the next, it is written as the
    // floating-point value it is rather than as a whole number it only looks like.
    assertEquals(Double.valueOf(1e17), Json.parse(Json.write(Double.valueOf(1e17))));

    // Neither infinity is a number JSON carries, in either direction of the axis.
    assertThrows(JsonException.class, () -> Json.write(Double.valueOf(Double.POSITIVE_INFINITY)));
    assertThrows(JsonException.class, () -> Json.write(Double.valueOf(Double.NEGATIVE_INFINITY)));
  }

  @Test
  void aDocumentSpacedOutTheWayAPersonWritesOneIsRead() {
    // All four characters JSON counts as space, around every part of a document rather than in the
    // one place a case happened to put them.
    Object read = Json.parse(" \t\r\n{ \"a\" : [ 1 , 2 ] , \"b\" : { } } \t\r\n");

    Map<String, Object> members = Wire.asFields(read, "a document");

    assertEquals(List.of(Long.valueOf(1), Long.valueOf(2)), members.get("a"));
    assertEquals(Map.of(), members.get("b"));
  }

  @Test
  void anExponentIsReadWhicheverWayItIsSpelled() {
    // The characters a number is allowed to carry beyond its digits, each of which a reader that
    // stopped at the first one it did not know would cut a number in half at.
    assertEquals(Double.valueOf(1000), Json.parse("1E+3"));
    assertEquals(Double.valueOf(1000), Json.parse("1e3"));
    assertEquals(Double.valueOf(0.001), Json.parse("1E-3"));
    assertEquals(Double.valueOf(-1.5e-3), Json.parse("-1.5e-3"));
  }

  @Test
  void aDocumentThatIsNothingAtAllIsRefusedRatherThanReadAsEmpty() {
    JsonException refused = assertThrows(JsonException.class, () -> Json.parse(null));

    assertTrue(refused.getMessage().contains("no document"), refused.getMessage());
  }

  /** What refusing that said, for a case holding several ceilings to their own words. */
  private static String refusing(Executable written) {
    return assertThrows(JsonException.class, written).getMessage();
  }
}

File: clients/kotlin/test/kotlin/com/hook0/kotlin/JsonTest.kt
package com.hook0.kotlin

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.Timeout

/**
 * The reader this artefact carries instead of a dependency.
 *
 * Every ceiling is a refusal rather than a truncation, which is the whole reason a reader of our own
 * is defensible at all: one that quietly read half a document would be worse than the dependency it
 * was written to avoid.
 */
@Timeout(60)
class JsonTest {

  @Test
  fun aDocumentReadsBackAsWhatItSays() {
    val read = Json.parse("""{"a":1,"b":[true,null,"x"],"c":{"d":1.5}}""")

    assertEquals(
      linkedMapOf<String, Any?>(
        "a" to 1L,
        "b" to listOf(true, null, "x"),
        "c" to linkedMapOf<String, Any?>("d" to 1.5)
      ),
      read
    )
  }

  @Test
  fun aWholeNumberReadsAsOneAndAFractionalOneDoesNot() {
    assertEquals(1L, Json.parse("1"))
    assertEquals(1.5, Json.parse("1.5"))
    assertEquals(1.0e3, Json.parse("1e3"))
  }

  @Test
  fun theOrderMembersArrivedInIsTheOrderTheyAreWrittenBackIn() {
    val document = """{"z":1,"a":2,"m":3}"""

    assertEquals(document, Json.write(Json.parse(document)))
  }

  @Test
  fun aDocumentThatCarriesSomethingPastItsEndIsRefused() {
    val refused = assertThrows(JsonException::class.java) { Json.parse("{} {}") }

    assertTrue(refused.message?.contains("past its end") == true, refused.message)
  }

  @Test
  fun everyEscapeJsonDeclaresIsReadAndWrittenBack() {
    val read = Json.parse("\"\\\" \\\\ \\/ \\b \\f \\n \\r \\t \\u0041\"")

    assertEquals("\" \\ / \b \u000C \n \r \t A", read)
    assertEquals("\"\\\" \\\\ / \\b \\f \\n \\r \\t A\"", Json.write(read))
  }

  @Test
  fun aStringCarryingARawControlCharacterIsRefused() {
    val refused = assertThrows(JsonException::class.java) { Json.parse("\"a\u0001b\"") }

    assertTrue(refused.message?.contains("control character") == true, refused.message)
  }

  @Test
  fun aDocumentNestedDeeperThanTheBoundIsRefusedRatherThanTrimmed() {
    val deep = "[".repeat(Json.MAX_DEPTH + 2) + "]".repeat(Json.MAX_DEPTH + 2)

    val refused = assertThrows(JsonException::class.java) { Json.parse(deep) }

    assertTrue(refused.message?.contains("nests deeper") == true, refused.message)
  }

  @Test
  fun aValueJsonHasNoWayToCarryIsRefusedRatherThanPrinted() {
    val refused = assertThrows(JsonException::class.java) { Json.write(Any()) }

    assertTrue(refused.message?.contains("is not something JSON carries") == true, refused.message)
  }

  @Test
  fun aNumberJsonHasNoWayToCarryIsRefused() {
    for (number in listOf(Double.NaN, Double.POSITIVE_INFINITY, Double.NEGATIVE_INFINITY)) {
      assertThrows(JsonException::class.java) { Json.write(number) }
    }
  }

  @Test
  fun anObjectKeyedBySomethingThatIsNotAStringIsRefused() {
    val refused = assertThrows(JsonException::class.java) { Json.write(mapOf(1 to "a")) }

    assertTrue(refused.message?.contains("keyed by") == true, refused.message)
  }

  @Test
  fun aWholeNumberWrittenAsADoubleTravelsWithoutAFraction() {
    assertEquals("12", Json.write(12.0))
    assertEquals("1.5", Json.write(1.5))
  }

  @Test
  fun everyCeilingThisWriterNamesRefusesOnTheWayOutAsWellAsOnTheWayIn() {
    // Each bound is written down twice, once for reading and once for writing, and until now neither
    // half of the collection ceilings nor the writing half of any of them was held to it. A ceiling
    // enforced in one direction only is a client that will compose a document it would refuse to
    // read back.
    val members = HashMap<String, Any?>()
    val items = ArrayList<Any?>(Json.MAX_ITEMS + 1)
    val crowded = StringBuilder("{")
    val listed = StringBuilder("[")
    for (item in 0..Json.MAX_ITEMS) {
      members["m$item"] = item.toLong()
      items.add(item.toLong())
      crowded.append(if (item > 0) "," else "").append("\"m").append(item).append("\":1")
      listed.append(if (item > 0) "," else "").append('1')
    }
    var nested: Any? = 1L
    for (depth in 0..Json.MAX_DEPTH) {
      nested = listOf(nested)
    }
    val whole = (0 until 10).map { "x".repeat(900_000) }

    assertTrue(refusing { Json.write(members) }.contains("members, above"))
    assertTrue(refusing { Json.write(items) }.contains("items, above"))
    assertTrue(refusing { Json.write(nested) }.contains("nests deeper"))
    assertTrue(refusing { Json.write("x".repeat(Json.MAX_STRING_CHARS + 8)) }.contains("characters long"))
    assertTrue(refusing { Json.write(whole) }.contains("characters long, above"))

    assertTrue(refusing { Json.parse("$crowded}") }.contains("members read"))
    assertTrue(refusing { Json.parse("$listed]") }.contains("items read"))
    val longString = "[\"" + "x".repeat(Json.MAX_STRING_CHARS + 8) + "\"]"
    val longDocument = "\"" + "x".repeat(Json.MAX_DOCUMENT_CHARS) + "\""

    assertTrue(refusing { Json.parse(longString) }.contains("characters read"))
    assertTrue(refusing { Json.parse(longDocument) }.contains("characters long"))
  }

  @Test
  fun aStringThatStopsInsideAnEscapeIsRefusedRatherThanReadUpToThere() {
    // Three ways an escape can be cut short or wrong, and a literal spelled almost right. Each ends
    // the document somewhere the reader cannot know what was meant, and reading what came before as
    // the whole of it would hand a caller a value the sender never wrote.
    for (malformed in listOf("\"a\\", "\"a\\u00\"", "\"a\\uzzzz\"", "tru", "nul", "fals")) {
      assertThrows(JsonException::class.java, { Json.parse(malformed) }, "`$malformed` was read as a document")
    }
  }

  @Test
  fun aCollectionThatRunsOnWithoutASeparatorIsRefusedRatherThanReadUpToThere() {
    for (malformed in listOf("{\"a\":1 \"b\":2}", "[1 2]")) {
      assertThrows(JsonException::class.java, { Json.parse(malformed) }, "`$malformed` was read as a document")
    }
  }

  @Test
  fun everyItemOfAnArrayIsWrittenAndSeparatedFromTheOneBefore() {
    // A separator only appears between the second item and the first, so an array of one never shows
    // whether there is one at all — and everything written until now was an array of one or none.
    assertEquals("[]", Json.write(emptyList<Any?>()))
    assertEquals("[1]", Json.write(listOf(1L)))
    assertEquals("[1,2,3]", Json.write(listOf(1L, 2L, 3L)))
    assertEquals(listOf(1L, 2L, 3L), Json.parse(Json.write(listOf(1L, 2L, 3L))))
  }

  @Test
  fun aNumberWithNothingAfterThePointIsWrittenWithoutOneWhicheverWidthItArrivedAs() {
    // A single-width float is the same number to the API as a double-width one, and travels the same
    // way: as the shorter of `7` and `7.0`, which is what JSON says they both are.
    assertEquals("7", Json.write(7.0f))
    assertEquals("1.5", Json.write(1.5f))
    assertEquals("-7", Json.write(-7.0))

    // Past where a double tells whole numbers apart one from the next, it travels as the
    // floating-point value it is rather than as a whole number it only looks like.
    assertEquals(1e17, Json.parse(Json.write(1e17)))
  }

  @Test
  fun aStringCarryingAControlCharacterIsWrittenAsTheCodePointItIs() {
    // The reader refuses a raw control character in a document; the writer has to make sure it never
    // produces one, which is the other side of the same rule.
    assertEquals("\"a\\u0001b\"", Json.write("a\u0001b"))
    assertEquals("a\u0001b", Json.parse("\"a\\u0001b\""))
  }

  @Test
  fun aDocumentSpacedOutTheWayAPersonWritesOneIsRead() {
    val read = Json.parse(" \t\r\n{ \"a\" : [ 1 , 2 ] , \"b\" : { } } \t\r\n")

    assertEquals(mapOf("a" to listOf(1L, 2L), "b" to emptyMap<String, Any?>()), read)
  }

  /** What refusing that said, for a case holding several ceilings to their own words. */
  private fun refusing(written: () -> Unit): String =
    assertThrows(JsonException::class.java) { written() }.message ?: ""
}
